### ISSUE TITLE
idea: A way to reflect prop changes to other places

### ISSUE THREAD

**Issue by trusktr** [skatejs/skatejs#1473]
## Feature idea

It'd be convenient to be able to reflect prop changes to other places easily.

### Use case(s)

**Generic use case:**

For example, suppose a custom element has `el.subObject.foo` and we want to reflect changes from `el.foo` to `el.subObject.foo`.

At the moment, we can implement it manually in `updated`:

```js
updated() {
  this.subObject.foo = this.foo
}
```

Would it be worth it to add some sort of mapping feature to the `props` API to make this easy?

**My specific use case:**

The nice thing about the `updated` sample above is that in my case the class that creates my `subObject` instance has `withUpdate` applied onto it, so it reacts to the value change! This is awesome! 👌  

I have to make sure that `connectedCallback` and other methods exist on the class and that they are called so that `withUpdate` works, which is what my [`element-behaviors`](https://github.com/trusktr/element-behaviors) lib handles.

Now what I'd like to do is map specific props of a behavior-hosting element to certain behaviors that care about those props.

Let me play with it and see what works...

**Comment by trusktr** [comment]
**My specific use case:**

The nice thing about the `updated` sample above is that in my case the class that creates my `subObject` instance has `withUpdate` applied onto it, so it reacts to the value change! 👍 

I have to make sure that `connectedCallback` and other methods exist on the class and that they are called so that `withUpdate` works, which is what my [`element-behaviors`](https://github.com/trusktr/element-behaviors) lib handles. I was playing with this a few days ago, and it's _awesome_ that it just works!

Now what I'd like to do is map specific props to certain behaviors, so that behaviors are in charge of handling changes to specific props of the element.

What I can do in my case is implement this functionality in my base class or a mixin, taking advantage of `withUpdate`s call to `updated`, to try out some ideas on how to make the mapping.

I'll be back once I've got a PoC implemented...

**Comment by trusktr** [comment]
Alright, so turns out, that attribute changes work great: when an element's attribute changes, `attributeChangedCallback` gets fired on every behavior that observes to the specific attribute, so that works great; changing an attribute on an element causes all my observing behaviors to have their `withUpdate` methods called.

What doesn't work is instance props. Changing those on the element doesn't cause anything to happen on the behaviors. If `Object.observe` had become a thing, this would be sooooo easy; I see how useful it could be but when used correctly.

**Things I'm stumbling on:**

- the `props` are defined in the `observedAttributes` getter, not at runtime, but I think what I need to do is define the props at runtime depending on which behaviors are added or removed from the element, so that certain props can be observable based on the current set of behaviors.
  - Again, this would be just so simple if we had `Object.observe`, and behaviors could use it on the `element`.
- I could do something similar to my [context idea](https://codepen.io/trusktr/project/editor/XbEOMk), but in that implementation we know ahead of time which props will change.
  - Behaviors can be arbitrarily added or removed from an element at runtime (that's powerful), which means we can't foresee which props will need to be observed ahead of time.
    - `Object.observe`. Aaaaaaargh! (I suppose a dynamic getter/setter implementation will have to do.)
- My elements with behaviors have a `this.behaviors` getter. I could loop through behaviors in `updated`, and forward props, but again this means I have to know ahead of time which props to define ahead of time in `static props` for it to work, and we don't know ahead of time which behaviors an element may have.
  - Object.observe!

TLDR:

- I don't think I personally need a forward-props feature with the current with-update implementation because I didn't realize at first that I wouldn't know which props to forward ahead of time. 
- For people that know ahead of time which props they'd like to forward and to where (pre-meditated), this might still be a nice feature.
- If we could define (and undefine?) props at runtime, then this feature could make more sense: I could define/undefine props as needed based on the current set of element behaviors.
- I think I'll have better luck polyfilling a (hacky) `Object.observe` function that uses getters/setters, then I'll use it in my element behaviors so that they can observe arbitrary props of their host element.
  -  **And because they use `withUpdate`, the result will be magical! This is the benefit of keeping the APIs free of external DOM API usage, that the mixins can be used in more places.** ✨

**Comment by trusktr** [comment]
This is what I had to do to get it working in my case (property reflection to another object):

```js
import { props } from 'skatejs'

// ... code omitted ...

    static props = {
            opacity: props.number,
            color: props.number,
    },

    async updated(oldProps) {
        this.updateMaterial(oldProps, 'color')
        this.updateMaterial(oldProps, 'opacity')
        super.updated && super.updated()
    }

    updateMaterial(oldProps, propName) {
        if (oldProps[propName] !== this[propName]) {
            this.element.threeObject3d.material[propName] = this[propName]
            this.element._needsToBeRendered()
        }
    }

// ...
```

So I made a `updateMaterial` method to handle forwarding props to some place.

Would it make sense to have a shortcut for this in the `props` API or something?

**Comment by treshugart** [comment]
> Would it make sense to have a shortcut for this in the props API or something?

I don't think so at this point. What you've done there is pretty straight forward and easy to do.

**Comment by trusktr** [comment]
This is covered by issue #1493 and PR https://github.com/skatejs/skatejs/pull/1509, which gives users flexibility to do things like described in this issue and other things.

**Comment by trusktr** [comment]
@treshugart And I made a helper for my case that greatly simplifies mapping to other places even more!:

```js
import {withUpdate} from 'skatejs'

// the helper
const mapTo = (prop, subObj) => ({
    ...prop,
    coerce(val, key) { return this[subObj][key] = prop.coerce(val) },
    deserialize(val, key) { return this[subObj][key] = prop.deserialize(val) },
})

class Mesh extends withUpdate(HTMLElement) {
        static props = {
            'cast-shadow': mapTo(props.boolean, 'threeObject3d'),
            'receive-shadow': mapTo(props.boolean, 'threeObject3d'),
        }
}
```

And I really need to I can improve it with lodash:

```js
const mapTo = (prop, path) => ({
    ...prop,
    coerce(val, key) { return _.set(this, path, value) = prop.coerce(val) },
    deserialize(val, key) { return _.set(this, path, value) = prop.deserialize(val) },
})

// ...

            'cast-shadow': mapTo(props.boolean, 'some.deep.nested.path'),
```

``

It's niiiiice, thanks to #1509!

**Comment by treshugart** [comment]
Does the `changed` prop callback in `v6-trey` solve this use case?

**Comment by trusktr** [comment]
I don't see it, maybe you already removed it?

**Comment by treshugart** [comment]
I renamed it to `set()` but it's essentially the same thing, just a better name, especially with `get()`.
